drivers/i2c/i2c_tca9546a.c

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright (c) 2020 Innoseis BV
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT ti_tca9546a
+
+#include <zephyr.h>
+#include <device.h>
+#include <devicetree.h>
+#include <drivers/i2c.h>
+#include <logging/log.h>
+#include <stdint.h>
+
+LOG_MODULE_REGISTER(tca9546a, CONFIG_I2C_LOG_LEVEL);
+
+struct tca9546a_root_config {
+	const struct device *bus;
+	uint16_t slave_addr;
+};
+
+struct tca9546a_root_data {
+	struct k_mutex lock;
+	uint8_t select_reg;
+#ifdef CONFIG_PM_DEVICE
+	uint32_t power_state;
+#endif
+};
+
+struct tca9546a_channel_config {
+	const struct device *root;
+	uint8_t addr;
+};
+
+static inline struct tca9546a_root_data *
+get_root_data_from_channel(const struct device *dev)
+{
+	const struct tca9546a_channel_config *channel_config = dev->config;
+
+	return channel_config->root->data;
+}
+
+static inline const struct tca9546a_root_config *
+get_root_config_from_channel(const struct device *dev)
+{
+	const struct tca9546a_channel_config *channel_config = dev->config;
+
+	return channel_config->root->config;
+}
+
+static int tca9546a_configure(const struct device *dev, uint32_t dev_config)
+{
+	const struct tca9546a_root_config *cfg = get_root_config_from_channel(
+		dev);
+
+	return i2c_configure(cfg->bus, dev_config);
+}
+
+static int tca9546a_set_selection(const struct device *dev, uint8_t sel)
+{
+	int res = 0;
+	struct tca9546a_root_data *data = dev->data;
+	const struct tca9546a_root_config *cfg = dev->config;
+
+	if (data->select_reg != sel) {
+		res = i2c_write(cfg->bus, &sel, 1, cfg->slave_addr);
+		if (res == 0) {
+			data->select_reg = sel;
+		}
+	}
+	return res;
+}
+
+static int tca9546a_transfer(const struct device *dev,
+			     struct i2c_msg *msgs,
+			     uint8_t num_msgs,
+			     uint16_t addr)
+{
+	struct tca9546a_root_data *data = get_root_data_from_channel(dev);
+	const struct tca9546a_root_config *config = get_root_config_from_channel(
+		dev);
+	const struct tca9546a_channel_config *down_cfg = dev->config;
+	int res;
+
+	res = k_mutex_lock(&data->lock, K_MSEC(5000));
+	if (res != 0) {
+		return res;
+	}
+
+	res = tca9546a_set_selection(down_cfg->root, down_cfg->addr);
+	if (res != 0) {
+		goto end_trans;
+	}
+
+	res = i2c_transfer(config->bus, msgs, num_msgs, addr);
+
+end_trans:
+	k_mutex_unlock(&data->lock);
+	return res;
+}
+
+static int tca9546_root_init(const struct device *dev)
+{
+	struct tca9546a_root_data *i2c_tca9546a = dev->data;
+	const struct tca9546a_root_config *config = dev->config;
+
+	if (!device_is_ready(config->bus)) {
+		LOG_ERR("I2C bus %s not ready", config->bus->name);
+		return -ENODEV;
+	}
+
+	i2c_tca9546a->select_reg = 0;
```

Comment:
Yes because `data->select_reg` is a bit field. So `tca9546a_set_selection` will never be called with '0' as parameter. So on the first transaction after initialization the transaction to set the hardware register will always happen.

Reply:
ok for me